### PR TITLE
Finish ValSet Pref Unbonding

### PR DESCRIPTION
**Pull request by mattverse**
Closes: #XXX

## What is the purpose of the change
Finishes the val set pref unbonding PR, mainly adds test cases, minor refactoring and better code documentation included. 

Note that this PR is pointed towards  `dev/valset-pref-unbond-issues` branch, not the main branch
## Testing and Verifying

This change added tests

## Documentation and Release Note

  - [ ] Does this pull request introduce a new feature or user-facing behavior changes?
  - [ ] Changelog entry added to `Unreleased` section of `CHANGELOG.md`?

Where is the change documented? 
  - [ ] Specification (`x/{module}/README.md`)
  - [ ] Osmosis documentation site
  - [ ] Code comments?
  - [ ] N/A

### CODE REVIEW

File: x/valset-pref/validator_set.go [hunk 1]
```diff
@@ -195,7 +193,6 @@ func (k Keeper) UndelegateFromValidatorSet(ctx sdk.Context, delegatorAddr string
 
 	// Step 6
```
Comment (czarcas7ic) [inline x/valset-pref/validator_set.go:194]: Can we give a synopsis of Step 6 and Step 7 in the comments like the others? Was nice/helpful

File: x/valset-pref/validator_set.go [hunk 2]
```diff
@@ -229,43 +223,46 @@ func (k Keeper) UndelegateFromValidatorSet(ctx sdk.Context, delegatorAddr string
...
-	var valSetRatio []ValRatio
+	var valSetRatios []ValRatio
 	validators := map[string]stakingtypes.Validator{}
 
 	for _, val := range prefs {
 		amountToUnDelegate := val.Weight.MulInt(undelegateAmt).TruncateInt()
 		valAddr, validator, err := k.getValAddrAndVal(ctx, val.ValOperAddress)
 		if err != nil {
-			return valSetRatio, validators, totalDelegatedAmt, err
+			return []ValRatio{}, map[string]stakingtypes.Validator{}, sdk.ZeroDec(), err
```
Comment (czarcas7ic) [inline x/valset-pref/validator_set.go:239]: question, does []ValRatio{} not implement nil?

File: x/valset-pref/validator_set_test.go [hunk 3]
```diff
@@ -147,6 +145,244 @@ func (s *KeeperTestSuite) TestIsValidatorSetEqual() {
...
 }
 
+func (s *KeeperTestSuite) TestUndelegateFromValidatorSet() {
+	tests := []struct {
+		name                  string
+		delegateAmt           []osmomath.Int
+		undelegateAmt         osmomath.Int
+		noValset              bool
+		expectedUndelegateAmt []osmomath.Int
+		expectedError         bool
```
Comment (czarcas7ic) [inline x/valset-pref/validator_set_test.go:155]: can we switch this from a bool to an error and match the error with ErrorIs

File: x/valset-pref/validator_set_test.go [hunk 4]
```diff
@@ -147,6 +145,244 @@ func (s *KeeperTestSuite) TestIsValidatorSetEqual() {
...
+	}{
+		{
```
Comment (czarcas7ic) [inline x/valset-pref/validator_set_test.go:157]: Isn't there another logic branch of exiting with Vratio of 1 that we could test here?

File: x/valset-pref/validator_set_test.go [hunk 5]
```diff
@@ -147,6 +145,244 @@ func (s *KeeperTestSuite) TestIsValidatorSetEqual() {
...
+					{
+						ValOperAddress: valAddrs[1],
+						Weight:         sdk.MustNewDecFromStr("0.666666666666666667"),
+					},
+				}
+			}
+
+			// set up delegation for each of the valset prefs
+			expectedTotalDelegatedAmt := sdk.ZeroDec()
+			if !test.notDelegated {
```
Comment (czarcas7ic) [inline x/valset-pref/validator_set_test.go:346]: double negative hurts my brain lol

**Review by czarcas7ic — APPROVED** [review]
Going to approve since this is just merging into another branch, but would like to review the branch being merged into main with more scrutiny.

File: x/valset-pref/validator_set.go (hunk 1 above)
Comment (mattverse) [reply x/valset-pref/validator_set.go:194]: I mentioned this on slack, but I personally don't understand if this part of the algorithm is even reachable, thus should be deleted. Will hold this comment for now and either add more comment or delete this once I gain context

File: x/valset-pref/validator_set.go (hunk 2 above)
Comment (mattverse) [reply x/valset-pref/validator_set.go:239]: Good catch! It does

File: x/valset-pref/validator_set_test.go (hunk 4 above)
Comment (mattverse) [reply x/valset-pref/validator_set_test.go:157]: Both of VRatio 1 and VRatio under 1 would go under same logic branch

File: x/valset-pref/validator_set_test.go (hunk 5 above)
Comment (mattverse) [reply x/valset-pref/validator_set_test.go:346]: Haha, If we use negative fields we can avoid having field set to true for all the other cases

File: x/valset-pref/validator_set_test.go (hunk 4 above)
Comment (mattverse) [reply x/valset-pref/validator_set_test.go:157]: Both of VRatio 1 and VRatio under 1 would go under same logic branch

**Comment by czarcas7ic** [comment]
Merging this, but going to transfer the comments about not reaching steps 5-7 / not testing them either, and will be a blocker for final merge. Thanks for getting this done!

**Pull request merged**